src/model/openthaigpt_pretraining_model/llama/model.py

```diff
@@ -1,14 +1,244 @@
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+
+        self.head_dim = args.dim // args.n_heads
+
+        self.n_local_heads = args.n_heads
+
+        self.wq = nn.Linear(
+            args.dim,
+            args.n_heads * self.head_dim,
+            bias=False,
+        )
+        self.wk = nn.Linear(
+            args.dim,
+            args.n_heads * self.head_dim,
+            bias=False,
+        )
+        self.wv = nn.Linear(
+            args.dim,
+            args.n_heads * self.head_dim,
+            bias=False,
+        )
+        self.wo = nn.Linear(
+            args.n_heads * self.head_dim,
+            args.dim,
+            bias=False,
+        )
+
+        self.cache_k = torch.zeros(
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        ).cuda()
+        self.cache_v = torch.zeros(
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        ).cuda()
+
+        if not (args.attention_mode == 'origin' or args.attention_mode == 'pytorch' or args.attention_mode == 'xformer'):
+            raise KeyError('attention mode must be "origin", "pytorch" or "xformer"')
+        
+        self.mode = args.attention_mode
+
+    def forward(self, x: torch.Tensor, start_pos: int, freqs_cis: torch.Tensor, mask: Optional[torch.Tensor]):
+        bsz, seqlen, _ = x.shape
+        xq, xk, xv = self.wq(x), self.wk(x), self.wv(x)
+
+        xq = xq.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+        xk = xk.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+        xv = xv.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+
+        xq, xk = apply_rotary_emb(xq, xk, freqs_cis=freqs_cis)
+
+        self.cache_k = self.cache_k.to(xq)
+        self.cache_v = self.cache_v.to(xq)
+
+        self.cache_k[:bsz, start_pos : start_pos + seqlen] = xk
+        self.cache_v[:bsz, start_pos : start_pos + seqlen] = xv
+
+        keys = self.cache_k[:bsz, : start_pos + seqlen]
+        values = self.cache_v[:bsz, : start_pos + seqlen]
+
+        xq = xq.transpose(1, 2)
+        keys = keys.transpose(1, 2)
+        values = values.transpose(1, 2)
+        
+        if self.mode == 'pytorch':
+            output = torch_attention(xq, keys, values, mask)
+        elif self.mode == 'xformer':
+            output = xformer_attention(xq, keys, values, mask)
+        elif self.mode == 'origin':
+            scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(self.head_dim)
+            if mask is not None:
+                scores = scores + mask  # (bs, n_local_heads, slen, cache_len + slen)
+            scores = F.softmax(scores.float(), dim=-1).type_as(xq)
+            output = torch.matmul(scores, values)  # (bs, n_local_heads, slen, head_dim)
+        
+        output = output.transpose(
+            1, 2
+        ).contiguous().view(bsz, seqlen, -1)
+
+        return self.wo(output)
+
+
+class FeedForward(nn.Module):
```

Comment:
This `FeedForward` model you chenged only `self.w1`, `self.w2`, and `self.w3` from `ColumnParallelLinear` and `RowParallelLinear` to `nn.Linear` but no changed in forward function. I suggest you to inherit `FeedForward` from original implementation.

This is example of my code.
```python
from llama.model import FeedForward as OriginalLlamaFeedForward

class FeedForward(OriginalLlamaFeedForward):
    def __init__(
        self,
        dim: int,
        hidden_dim: int,
        multiple_of: int,
    ):
        super().__init__(dim, hidden_dim, multiple_of)
        self.w1 = nn.Linear(
            dim, hidden_dim, bias=False
        )
        self.w2 = nn.Linear(
            hidden_dim, dim, bias=False
        )
        self.w3 = nn.Linear(
            dim, hidden_dim, bias=False
        )
```

Reply:
If use this way we will set up fairscale that useless

Reply:
How?

Reply:
How long setup time?

Reply:
I have never setup and fairscale @new5558 called will don't use in this project